File: training-data/haskell/Isbn10Validator.hs
import Data.Char (digitToInt, isDigit)

isValidIsbn10 :: String -> Bool
isValidIsbn10 isbn = case filter (/= '-') isbn of
  s | length s == 10, all isDigit (init s), isDigit (last s) || last s == 'X' ->
        sum (zipWith (*) [10, 9 .. 1] (map value s)) `mod` 11 == 0
  _ -> False
  where
    value 'X' = 10
    value c = digitToInt c

main :: IO ()
main = mapM_ (\s -> putStrLn (s ++ " " ++ show (isValidIsbn10 s))) ["3-598-21508-8", "3-598-21507-X", "3-598-21508-9"]

File: training-data/kotlin/Isbn10Validator.kt
fun isValidIsbn10(isbn: String): Boolean {
    val s = isbn.replace("-", "")
    if (!Regex("""\d{9}[\dX]""").matches(s)) return false
    return s.withIndex().sumOf { (i, c) -> (if (c == 'X') 10 else c.digitToInt()) * (10 - i) } % 11 == 0
}

fun main() {
    listOf("3-598-21508-8", "3-598-21507-X", "3-598-21508-9").forEach { println("$it ${isValidIsbn10(it)}") }
}
